### PR TITLE
[cmd setup] Add and deprecate setup cmds for index handling

### PR DESCRIPTION
**Pull request by simitt**
Deprecate `setup --template` and `setup --ilm` in favor of 
newly introduced `setup --index-management` command.

implements #12095

### CODE REVIEW

File: libbeat/idxmgmt/std.go [hunk 1]
```diff
@@ -62,6 +62,33 @@ type ilmIndexSelector struct {
 	st    *indexState
 }
 
+type componentName uint8
+
+//go:generate stringer -linecomment -type componentName
+const (
+	componentTemplate componentName = iota //template
+	componentILM                           //ilm
+)
```
Comment (urso) [inline libbeat/idxmgmt/std.go]: would `kind` or `type` be a better match here?

File: libbeat/idxmgmt/std.go [hunk 2]
```diff
@@ -176,41 +213,55 @@ func (s *indexSupport) BuildSelector(cfg *common.Config) (outputs.IndexSelector,
 		st:    &s.st,
 	}, nil
 }
+func (m *indexManager) VerifySetup(loadTemplate, loadILM LoadMode) (bool, string) {
```
Comment (urso) [inline libbeat/idxmgmt/std.go]: nit: newlines between longer functions.

File: libbeat/idxmgmt/std.go [hunk 3]
```diff
@@ -176,41 +213,55 @@ func (s *indexSupport) BuildSelector(cfg *common.Config) (outputs.IndexSelector,
...
+			"to issues and is not recommended. Check your configuration"
+	}
+	var warn string
+	if !ilmComponent.load {
+		warn += "ILM policy and write alias loading not enabled. "
+	}
+	if !templateComponent.load {
+		warn += "Template loading not enabled."
+	}
+	return warn == "", warn
```
Comment (urso) [inline libbeat/idxmgmt/std.go]: Consider error as second return value.

File: libbeat/idxmgmt/std.go [hunk 4]
```diff
@@ -62,6 +62,33 @@ type ilmIndexSelector struct {
...
+//go:generate stringer -linecomment -type componentName
+const (
+	componentTemplate componentName = iota //template
+	componentILM                           //ilm
+)
+
+type component struct {
+	name                     componentName
+	enabled, overwrite, load bool
+}
```
Comment (urso) [inline libbeat/idxmgmt/std.go]: I'm wondering if 'capability' or 'feature' might be a better name.

File: libbeat/idxmgmt/std.go [hunk 5]
```diff
@@ -254,6 +298,22 @@ func (m *indexManager) Setup(loadTemplate, loadILM LoadMode) error {
...
+	var err error
+	withILM := m.support.st.withILM.Load()
+	if !withILM {
+		withILM, err = m.ilm.Enabled()
+		if err != nil {
+			return false, err
+		}
+		if withILM {
+			// mark ILM as enabled in indexState
+			m.support.st.withILM.CAS(false, true)
```
Comment (urso) [inline libbeat/idxmgmt/std.go]: Where is the timestamp set?

**Comment by urso** [comment]
PR seems to do 2 things:
1. Add and deprecate flags
2. introduce 'component' type for internal refactoring purposes.

Why is the refactoring needed?

File: libbeat/idxmgmt/std.go (hunk 5 above)
Comment (simitt) [reply libbeat/idxmgmt/std.go]: Not sure what you mean. I moved it from https://github.com/elastic/beats/pull/12132/files#diff-a7b20900159263d87105489854d4d628L202, related to https://github.com/elastic/beats/pull/12132/files#diff-a7b20900159263d87105489854d4d628L184

File: libbeat/idxmgmt/std.go (hunk 3 above)
Comment (simitt) [reply libbeat/idxmgmt/std.go]: I didn't want to indicate it is an error, as the combinations are still valid, but we should print warnings.

**Comment by simitt** [comment]
I refactored because the checks with all the different load modes together with `enabled` and `overwrite` was getting very complex already and not easy to follow.

File: libbeat/idxmgmt/std.go (hunk 5 above)
Comment (urso) [reply libbeat/idxmgmt/std.go]: Nevermind, I confused it with some other code.

File: libbeat/idxmgmt/std_test.go [hunk 6]
```diff
@@ -197,7 +215,59 @@ func TestDefaultSupport_BuildSelector(t *testing.T) {
...
+				"setup.template.enabled": setup.tmpl,
+			})
+			require.NoError(t, err)
+			support, err := MakeDefaultSupport(ilm.StdSupport)(nil, beat.Info{}, cfg)
+			require.NoError(t, err)
+			manager := support.Manager(newMockClientHandler(), nil)
+			ok, warn := manager.VerifySetup(setup.loadTmpl, setup.loadILM)
+			assert.Equal(t, setup.ok, ok)
+			assert.Contains(t, warn, setup.warn)
+		})
```
Comment (urso) [inline libbeat/idxmgmt/std_test.go]: when using the mockClientHandler always run `assertInvariants` at the end of an test. E.g. for `VerifySetup` no ops should have been recorded (adds some behavioral testing).

File: libbeat/tests/system/test_cmd_setup_index_management.py [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+
+        self.es = self.es_client()
+        self.idxmgmt = IdxMgmt(self.es, self.index_name)
+        self.idxmgmt.delete(indices=[self.custom_alias, self.custom_policy, self.index_name])
+
+        logging.getLogger("urllib3").setLevel(logging.WARNING)
+        logging.getLogger("elasticsearch").setLevel(logging.ERROR)
+
+    def tearDown(self):
+        self.idxmgmt.delete(indices=[self.custom_alias, self.custom_policy, self.index_name])
```
Comment (urso) [inline libbeat/tests/system/test_cmd_setup_index_management.py]: is order important here? Maybe put policy name last always.

**Comment by simitt** [comment]
jenkins, test this

**Comment by simitt** [comment]
jenkins, test this

**Pull request merged**